wallet/storage/src/internal/mod.rs

```diff
@@ -31,19 +36,99 @@ use wallet_types::{
     WalletTx,
 };
 
+use self::store_tx::EncryptionState;
+
 /// Store for wallet data, parametrized over the backend B
-pub struct Store<B: storage::Backend>(storage::Storage<B, Schema>);
+pub struct Store<B: storage::Backend> {
+    storage: storage::Storage<B, Schema>,
+    encryption_state: EncryptionState,
+}
```

Comment:
Probably more for the future but wondering whether it would make sense to keep track of the encryption state at compilation time.

Rough sketch:

```rust
struct Locked;
struct Unlocked(Option<Key>);

struct Store<B, ES> { storage: Storage<B>, encryption: ES }
impl Store { fn new() -> Result<Either<Store<B, Locked>, Store<B, Unlocked>>>; }
impl Store<B, Locked> {
    fn unlock(self, passwd) -> Result<Store<B, Unlocked>>;
}
impl Store<B, Unlocked> {
    fn lock(self) -> Store<B, Locked>;
    // Other operations only applicable to unlocked storage
}
```

Thoughts? Is the usage too dyanmic for something like that to be useful?

Reply:
> This is nice, and is a common pattern in Rust. The problem is that the call decisions are dynamic from the user. There are wallet commands at the controller level to lock/unlock that eventually yield these calls at the low level.

You can also have the dynamic version by introducing something like `MaybeLocked`:
```rust
enum MaybeLocked { Locked | Unlocked(Option<Key>) }
impl Store<B, MaybeLocked> { fn check(self) -> Either<Store<B, Locked>, Store<B, Unlocked>>; }
```

That way:
* If the storage is `MaybeLocked`, you are forced to check.
* If you know it's unlocked, you can access it without checking.

Not suggesting this is a good idea, it all depends how is the store used. Just pointing out something like that is possible.

Reply:
The Store is always decided at runtime if it is locked or unlocked. But I can modify the transaction objects to have separate types for `ReadOnlyLocked`, `ReadOnlyUnlocked`, `ReadWriteLocked`, `ReadWriteUnlocked`
where the Locked variants will not have any methods to operate on the encrypted things (private keys)
and the Unlocked variants will have them.

I also need to separate the actual locking/unlocking/encrypting methods to a private trait so that others outside the Store can't change the state through a transaction object.